betty/model/association.py

```diff
@@ -5,35 +5,129 @@
 from __future__ import annotations
 
 import weakref
-from abc import abstractmethod
-from typing import Generic, cast, Any, Iterable, TypeVar, final
+from abc import abstractmethod, ABC
+from typing import (
+    Generic,
+    cast,
+    Any,
+    Iterable,
+    TypeVar,
+    final,
+    TypeAlias,
+    overload,
+    Never,
+)
 
 from basedtyping import Intersection
 from typing_extensions import override
 
-from betty.attr import DeletableAttr
+from betty.attr import DeletableAttr, SettableAttr
 from betty.classtools import repr_instance
 from betty.importlib import import_any
 from betty.model import Entity
 from betty.model.collections import EntityCollection, SingleTypeEntityCollection
-from betty.typing import internal
 
+_T = TypeVar("_T")
 _EntityT = TypeVar("_EntityT", bound=Entity)
 _OwnerT = TypeVar("_OwnerT")
 _AssociateT = TypeVar("_AssociateT")
 _AssociationAttrValueT = TypeVar("_AssociationAttrValueT")
 _AssociationAttrSetT = TypeVar("_AssociationAttrSetT")
 
 
-class Association(
+class AssociationRequired(RuntimeError):
+    """
+    Raised when an operation cannot be performed because the association in question is required.
+
+    These are preventable by checking :py:attr:`betty.model.association.Association.required`.
+    """
+
+    pass
+
+
+class ResolutionError(RuntimeError):
+    """
+    Raised when a :py:class:`betty.model.association.Resolver` cannot successfully resolve itself.
+    """
+
+    pass
+
+
+class _Resolver(Generic[_T], ABC):
+    @abstractmethod
+    def resolve(self) -> _T:
+        """
+        Return the resolved entity or entities.
+
+        :raises ResolutionError: Raised if resolution failed.
+        """
+        pass
+
+
+class RequiredEntityResolver(Generic[_EntityT], _Resolver[_EntityT]):
```

Comment:
Do we really need the required/optional distinction here? If we return Entity it will satisfy any Entity|None.